test/test_diag.jl

```diff
@@ -0,0 +1,54 @@
+# Points
+x0 = [-1.0, 1.0, -1.0]
+x1 = x0 + [1.0, 1.0, 1.0]
+
+# Test functions
+# f(x) = x[1]^2 + x[2]^2 + x[3]^2
+∇f(x) = 2 * [x[1], x[2], x[3]]
+
+# g(x) = exp(x[1]) + x[2] + cos(x[3])
+∇g(x) = [exp(x[1]), 1, -sin(x[3])]
+
+# h(x) = x[1]^2 * x[2] * x[3]^3
+∇h(x) = [2 * x[1] * x[2] * x[3]^3, x[1]^2 * x[3]^3, 3 * x[1]^2 * x[2] * x[3]^2]
+
+@testset "Weak secant equation" begin
+  for grad_fun in (:∇f, :∇g, ∇h)
+    grad = eval(grad_fun)
+    s = x1 - x0
+    y = grad(x1) - grad(x0)
+    B = DiagonalQN([1.0, -1.0, 1.0])
+    push!(B, s, y)
+    @test abs(dot(s, B * s) - dot(s, y)) <= 1e-10
```

Comment:
If you want, you could add a test to verify that `@allocated mul!(res, B, s) == 0` (you will have to call `mul!` 1 time before for compilation), but this is not mandatory.